modules/sdk/src/main/scala/org/tessellation/sdk/domain/cluster/programs/Joining.scala

```diff
@@ -152,14 +152,17 @@ sealed abstract class Joining[F[_]: Async: GenUUID: SecurityProvider: KryoSerial
   private def validateHandshakeConditions(toPeer: PeerToJoin): F[Unit] =
     validateHandshakeConditions(toPeer.id, toPeer.ip, toPeer.p2pPort)
 
-  def joinRequest(joinRequest: JoinRequest, remoteAddress: Host): F[Unit] =
+  def joinRequest(hasCollateral: PeerId => F[Boolean])(joinRequest: JoinRequest, remoteAddress: Host): F[Unit] =
```

Comment:
As we discussed, because `Joining` is initialized in `sdk` and  snapshot storages are initialized on `core` or `dag-l1` level this is the way to pass verification function to `Joining`.